tests/accessor/accessor_common.h

```diff
@@ -864,6 +864,105 @@ decltype(auto) get_subscript_overload(const AccT& accessor, size_t index) {
+    if (!empty) CHECK(value_operations::are_equal(*(--it), exp_last));
+  }
+  {
+    INFO("check cbegin() method");
+    auto it = accessor.cbegin();
+    STATIC_CHECK(std::is_same_v<decltype(it), typename AccT::const_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*it, exp_first));
+  }
+  {
+    INFO("check cend() method");
+    auto it = accessor.cend();
+    STATIC_CHECK(std::is_same_v<decltype(it), typename AccT::const_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*(--it), exp_last));
+  }
+  {
+    INFO("check rbegin() method");
+    auto it = accessor.rbegin();
+    STATIC_CHECK(std::is_same_v<decltype(it), typename AccT::reverse_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*it, exp_last));
+  }
+  {
+    INFO("check rend() method");
+    auto it = accessor.rend();
+    STATIC_CHECK(std::is_same_v<decltype(it), typename AccT::reverse_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*(--it), exp_first));
+  }
+  {
+    INFO("check crbegin() method");
+    auto it = accessor.crbegin();
+    STATIC_CHECK(
+        std::is_same_v<decltype(it), typename AccT::const_reverse_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*it, exp_last));
+  }
+  {
+    INFO("check crend() method");
+    auto it = accessor.crend();
+    STATIC_CHECK(
+        std::is_same_v<decltype(it), typename AccT::const_reverse_iterator>);
+    if (!empty) CHECK(value_operations::are_equal(*(--it), exp_first));
+  }
+}
+
+template <typename AccT, typename T = int>
+bool test_begin_end_device(AccT& accessor, T exp_first = T(), T exp_last = T(),
+                           bool empty = true) {
+  bool res = true;
+
+  auto it_begin = accessor.begin();
+  res &= std::is_same_v<decltype(it_begin), typename AccT::iterator>;
+  auto it_end = accessor.end();
+  res &= std::is_same_v<decltype(it_end), typename AccT::iterator>;
+
+  auto it_cbegin = accessor.cbegin();
+  res &= std::is_same_v<decltype(it_cbegin), typename AccT::const_iterator>;
+
+  auto it_cend = accessor.cend();
+  res &= std::is_same_v<decltype(it_cend), typename AccT::const_iterator>;
+
+  auto it_rbegin = accessor.rbegin();
+  res &= std::is_same_v<decltype(it_rbegin), typename AccT::reverse_iterator>;
+
+  auto it_rend = accessor.rend();
+  res &= std::is_same_v<decltype(it_rend), typename AccT::reverse_iterator>;
+
+  auto it_crbegin = accessor.crbegin();
+  res &= std::is_same_v<decltype(it_crbegin),
+                        typename AccT::const_reverse_iterator>;
+
+  auto it_crend = accessor.crend();
+  res &=
+      std::is_same_v<decltype(it_crend), typename AccT::const_reverse_iterator>;
+
+  if (!empty) {
+    res &= value_operations::are_equal(*it_begin, exp_first);
+    res &= value_operations::are_equal(*(--it_end), exp_first);
+    res &= value_operations::are_equal(*it_cbegin, exp_first);
+    res &= value_operations::are_equal(*(--it_cend), exp_first);
```

Comment:
```suggestion
    res &= value_operations::are_equal(*it_begin, exp_first);
    res &= value_operations::are_equal(*(--it_end), exp_last);
    res &= value_operations::are_equal(*it_cbegin, exp_first);
    res &= value_operations::are_equal(*(--it_cend), exp_last);
```
Please, increase the size of the buffer to two elements and use different element values to catch bugs like this.

Reply:
> Please, increase the size of the buffer to two elements and use different element values to catch bugs like this.

Fixed. This check is used for buffers with different sizes.

Reply:
> Please disregard erroneous commit message, it doesn't change buffer sizes, buffer with different sizes were already used.

Did you run tests? Do they fail on these checks?

Reply:
> Did you run tests? Do they fail on these checks?
Currently this test can't be run because not all functionality that are checked here is implemented.